lib/puppet/functions/vault_lookup/lookup_key.rb

```diff
@@ -0,0 +1,73 @@
+Puppet::Functions.create_function(:'vault_lookup::lookup_key') do
+  dispatch :lookup_key do
+    param 'String', :path
+    param 'String', :key
+    optional_param 'String', :vault_url
+  end
+
+  def lookup_key(path, key, vault_url = nil)
+    if vault_url.nil?
+      Puppet.debug 'No Vault address was set on function, defaulting to value from VAULT_ADDR env value'
+      vault_url = ENV['VAULT_ADDR']
+      raise Puppet::Error, 'No vault_url given and VAULT_ADDR env variable not set' if vault_url.nil?
+    end
+
+    uri = URI(vault_url)
+    # URI is used here to just parse the vault_url into a host string
+    # and port; it's possible to generate a URI::Generic when a scheme
+    # is not defined, so double check here to make sure at least
+    # host is defined.
+    raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless uri.hostname
```

Comment:
Unless this accounts for my above concerns with an empty string.

Reply:
It does in that it freaks out if it's not a valid url (and an empty string qualifies) -- it's not really the cleanest implementation.  That said it mirrors what is (or was if it's changed) in lookup.rb  =)  I somewhat wonder if this would be better served as an additional optional argument to lookup itself.  However it was a quick solution and I'm using it in production so it's got that goin' for it.  =D

Going to add another comment related to this kinda.